cmd/server/artifacts/artifact_server.go

```diff
@@ -0,0 +1,183 @@
+package artifacts
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/argoproj/argo/cmd/server/auth"
+	"github.com/argoproj/argo/persist/sqldb"
+	wfv1 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+	artifact "github.com/argoproj/argo/workflow/artifacts"
+	"github.com/argoproj/argo/workflow/packer"
+)
+
+type ArtifactServer struct {
+	authN                 auth.Gatekeeper
+	offloadNodeStatusRepo sqldb.OffloadNodeStatusRepo
+	wfArchive             sqldb.WorkflowArchive
+}
+
+func NewArtifactServer(authN auth.Gatekeeper, offloadNodeStatusRepo sqldb.OffloadNodeStatusRepo, wfArchive sqldb.WorkflowArchive) *ArtifactServer {
+	return &ArtifactServer{authN, offloadNodeStatusRepo, wfArchive}
+}
+
+func (a *ArtifactServer) GetArtifact(w http.ResponseWriter, r *http.Request) {
+
+	ctx, err := a.gateKeeping(r)
+	if err != nil {
+		w.WriteHeader(401)
+		_, _ = w.Write([]byte(err.Error()))
+		return
+	}
+	path := strings.SplitN(r.URL.Path, "/", 6)
+
+	namespace := path[2]
+	workflowName := path[3]
+	nodeId := path[4]
+	artifactName := path[5]
+
+	log.WithFields(log.Fields{"namespace": namespace, "workflowName": workflowName, "nodeId": nodeId, "artifactName": artifactName}).Info("Download artifact")
+
+	wf, err := a.getWorkflow(ctx, namespace, workflowName)
+	if err != nil {
+		a.serverInternalError(err, w)
+		return
+	}
+
+	data, err := a.getArtifact(ctx, wf, nodeId, artifactName)
+	if err != nil {
+		a.serverInternalError(err, w)
+		return
+	}
+	a.ok(w, data)
+}
+func (a *ArtifactServer) GetArtifactByUID(w http.ResponseWriter, r *http.Request) {
+
+	ctx, err := a.gateKeeping(r)
+	if err != nil {
+		w.WriteHeader(401)
+		_, _ = w.Write([]byte(err.Error()))
+		return
+	}
+	path := strings.SplitN(r.URL.Path, "/", 6)
+
+	uid := path[2]
+	nodeId := path[3]
+	artifactName := path[4]
+
+	log.WithFields(log.Fields{"uid": uid, "nodeId": nodeId, "artifactName": artifactName}).Info("Download artifact")
+
+	wf, err := a.getWorkflowByUID(ctx, uid)
+	if err != nil {
+		a.serverInternalError(err, w)
```

Comment:
we are clobbering the authz 403 error and always returning 500.

Reply:
Sure. I have an issue to revisit security for artifacts.